Zend/zend_execute.c

```diff
@@ -764,6 +787,128 @@ static int zend_verify_internal_arg_type(zend_function *zf, uint32_t arg_num, zv
+{
+	if (cache_slot && EXPECTED(Z_OBJCE_P(object) == CACHED_PTR_EX(cache_slot))) {
+		zend_property_info *info = (zend_property_info*)CACHED_PTR_EX(cache_slot + 2);
+
+		return (info && info->type) ? info : NULL;
+	}
+
+	do {
+		if (UNEXPECTED(Z_TYPE_P(property) != IS_STRING)) {
+			break;
+		}
+
+		if (UNEXPECTED(Z_TYPE_P(object) != IS_OBJECT)) {
+			if (Z_ISREF_P(object)) {
+				object = Z_REFVAL_P(object);
+				if (Z_TYPE_P(object) != IS_OBJECT)
+					break;
+			} else {
+				break;
+			}
+		}
+
+		return zend_object_fetch_property_type_info_ex(object, Z_STR_P(property), cache_slot);
+	} while (0);
+
+	return NULL;
+}
+
+static inline zend_bool zend_verify_scalar_property_type(zend_uchar type, zval *property, zend_bool strict) {
+	if (EXPECTED(!strict)) {
+		if (UNEXPECTED(!ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property)))) {
+			return zend_verify_weak_scalar_type_hint(type, property);
+		}
+		return 1;
+	}
+
+	if (UNEXPECTED(type == IS_DOUBLE && Z_TYPE_P(property) == IS_LONG)) {
+		double dest = (double) Z_LVAL_P(property);
+		zval_ptr_dtor(property);
+		ZVAL_DOUBLE(property, dest);
+		return 1;
+	}
+
+	return ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property));
+}
+
+void zend_verify_property_type_error(zend_property_info *info, zend_string *name, zval *property) {
+	zend_string *resolved = zend_resolve_property_type(info->type_name, info->ce);
+	
+	if (info->type == IS_OBJECT) {
+		zend_throw_exception_ex(zend_ce_type_error, info->type, 
+					"Typed property %s::$%s must be an instance of %s, %s used",
+						ZSTR_VAL(info->ce->name),
+						ZSTR_VAL(name),
+						ZSTR_VAL(resolved),
+						Z_TYPE_P(property) == IS_OBJECT ?
+							ZSTR_VAL(Z_OBJCE_P(property)->name) :
+							zend_get_type_by_const(Z_TYPE_P(property)));
+	} else {
+		zend_throw_exception_ex(zend_ce_type_error, info->type, 
+					"Typed property %s::$%s must be %s, %s used",
+						ZSTR_VAL(info->ce->name),
+						ZSTR_VAL(name),
+						zend_get_type_by_const(info->type),
+						Z_TYPE_P(property) == IS_OBJECT ?
+							ZSTR_VAL(Z_OBJCE_P(property)->name) :
+							zend_get_type_by_const(Z_TYPE_P(property)));
+	}
+}
+
+zend_bool zend_verify_property_type(zend_property_info *info, zval *property, zend_bool strict) {
+	switch (info->type) {
+		case IS_OBJECT: {
+			zend_string *resolved = zend_resolve_property_type(info->type_name, info->ce);
+			
+			if (!info->type_ce) {
+				info->type_ce = zend_lookup_class(resolved);
+			}
+
+			if (!info->type_ce || Z_TYPE_P(property) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(property), info->type_ce)) {
+				return 0;
+			}
+			return 1;
+		}
+
+		case IS_CALLABLE: switch (Z_TYPE_P(property)) {
+			case IS_OBJECT:
+				if (instanceof_function(zend_ce_closure, Z_OBJCE_P(property)))
+					return 1;
+
+			case IS_ARRAY:
+				if (zend_is_callable(property, IS_CALLABLE_CHECK_SILENT, NULL))
+					return 1;
+		}
+
+		default:
+			if (!zend_verify_scalar_property_type(info->type, property, strict)) {
+				return 0;
+			}
+			return 1;
+	}
+}
+
```

Comment:
This function should be optimized to check the most frequent valid condition first - "if (info->type == Z_TYPE_P(property)) return 1;" and only then check for other less frequent use cases.